// particle-api-js/lib/Particle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _superagentBinaryParser = require('./superagent-binary-parser');

var _superagentBinaryParser2 = _interopRequireDefault(_superagentBinaryParser);

var _Defaults = require('./Defaults');

var _Defaults2 = _interopRequireDefault(_Defaults);

var _EventStream = require('./EventStream');

var _EventStream2 = _interopRequireDefault(_EventStream);

var _Agent = require('./Agent');

var _Agent2 = _interopRequireDefault(_Agent);

var _Client = require('./Client');

var _Client2 = _interopRequireDefault(_Client);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Particle Cloud API wrapper.
 *
 * See <https://docs.particle.io/reference/javascript/> for examples
 * of using the `Particle` class.
 *
 * Most Particle methods take a single unnamed argument object documented as
 * `options` with key/value pairs for each option.
 */
var Particle = function () {
	/**
  * Contructor for the Cloud API wrapper.
  *
  * Create a new Particle object and call methods below on it.
  *
  * @param  {Object} options Options for this API call Options to be used for all requests (see [Defaults](../src/Defaults.js))
  */
	function Particle() {
		var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
		(0, _classCallCheck3.default)(this, Particle);

		// todo - this seems a bit dangerous - would be better to put all options/context in a contained object
		(0, _assign2.default)(this, _Defaults2.default, options);
		this.context = {};
		this.agent = new _Agent2.default(this.baseUrl);
	}

	(0, _createClass3.default)(Particle, [{
		key: '_isValidContext',
		value: function _isValidContext(name, context) {
			return (name === 'tool' || name === 'project') && context !== undefined;
		}
	}, {
		key: 'setContext',
		value: function setContext(name, context) {
			if (context !== undefined) {
				if (this._isValidContext(name, context)) {
					this.context[name] = context;
				} else {
					throw Error('uknown context name or undefined context: ' + name);
				}
			}
		}

		/**
   * Builds the final context from the context parameter and the context items in the api.
   * @param  {Object} context       The invocation context, this takes precedence over the local context.
   * @returns {Object} The context to use.
   * @private
   */

	}, {
		key: '_buildContext',
		value: function _buildContext(context) {
			return (0, _assign2.default)(this.context, context);
		}

		/**
   * Login to Particle Cloud using an existing Particle acccount.
   * @param  {Object} options Options for this API call
   * @param  {String} options.username      Username for the Particle account
   * @param  {String} options.password      Password for the Particle account
   * @param  {Number} options.tokenDuration How long the access token should last in seconds
   * @returns {Promise} A promise
   */

	}, {
		key: 'login',
		value: function login(_ref) {
			var username = _ref.username,
			    password = _ref.password,
			    _ref$tokenDuration = _ref.tokenDuration,
			    tokenDuration = _ref$tokenDuration === undefined ? this.tokenDuration : _ref$tokenDuration,
			    context = _ref.context;

			return this.request({ uri: '/oauth/token', form: {
					username: username,
					password: password,
					grant_type: 'password',
					client_id: this.clientId,
					client_secret: this.clientSecret,
					expires_in: tokenDuration
				}, method: 'post', context: context });
		}

		/**
   * If login failed with an 'mfa_required' error, this must be called with a valid OTP code to login
   * @param  {Object} options Options for this API call
   * @param  {String} options.mfaToken Given as 'mfa_token' in the error body of `.login()`.
   * @param  {String} options.otp      Current one-time-password generated from the authentication application
   * @param  {Number} options.context  Request context
   * @returns {Promise} A promise
   */

	}, {
		key: 'sendOtp',
		value: function sendOtp(_ref2) {
			var mfaToken = _ref2.mfaToken,
			    otp = _ref2.otp,
			    context = _ref2.context;

			return this.request({ uri: '/oauth/token', form: {
					grant_type: 'urn:custom:mfa-otp',
					mfa_token: mfaToken,
					otp: otp,
					client_id: this.clientId,
					client_secret: this.clientSecret
				}, method: 'post', context: context });
		}

		/**
   * Enable MFA on the currently logged in user
   * @param {Object} options	Options for this API call
   * @param {Object} options.auth		access token
   * @param {Object} options.context	Request context
   * @returns {Promise} A promise
   */

	}, {
		key: 'enableMfa',
		value: function enableMfa(_ref3) {
			var auth = _ref3.auth,
			    context = _ref3.context;

			return this.get('/v1/user/mfa-enable', auth, undefined, context);
		}

		/**
   * Confirm MFA for the user. This must be called with current TOTP code, determined from the results of enableMfa(). You will be prompted to enter an OTP code every time you login after enrollment is confirmed.
   * @param {Object} options	Options for this API call
   * @param {Object} options.auth		access token
   * @param {Object} options.mfaToken	Token given from previous step to
   * @param {Object} options.otp		Current one-time-password generated from the authentication app
   * @param {Object} options.context	Request context
   * @returns {Promise} A promise
   */

	}, {
		key: 'confirmMfa',
		value: function confirmMfa(_ref4) {
			var auth = _ref4.auth,
			    mfaToken = _ref4.mfaToken,
			    otp = _ref4.otp,
			    context = _ref4.context;

			return this.post('/v1/user/mfa-enable', { mfa_token: mfaToken, otp: otp }, auth, context);
		}

		/**
   * Disable MFA for the user.
   * @param {Object} options	Options for this API call
   * @param {Object} options.auth				access token
   * @param {Object} options.currentPassword	User's current password
   * @param {Object} options.context			Request context
   * @returns {Promise} A promise
   */

	}, {
		key: 'disableMfa',
		value: function disableMfa(_ref5) {
			var auth = _ref5.auth,
			    currentPassword = _ref5.currentPassword,
			    context = _ref5.context;

			return this.put('/v1/user/mfa-disable', { current_password: currentPassword }, auth, context);
		}

		/**
   * Create Customer for Product.
   * @param  {Object} options Options for this API call
   * @param  {String} options.email         Username for the Particle account
   * @param  {String} options.password      Password for the Particle account
   * @param  {String} options.product       Create the customer in this product ID or slug
   * @returns {Promise} A promise
   */

	}, {
		key: 'createCustomer',
		value: function createCustomer(_ref6) {
			var email = _ref6.email,
			    password = _ref6.password,
			    product = _ref6.product,
			    context = _ref6.context;

			var uri = '/v1/products/' + product + '/customers';
			return this.request({ uri: uri, form: {
					email: email,
					password: password,
					grant_type: 'client_credentials',
					client_id: this.clientId,
					client_secret: this.clientSecret
				}, method: 'post', context: context });
		}

		/**
   * Login to Particle Cloud using an OAuth client.
   * @param  {Object} options Options for this API call
   * @param  {Object} options.context   Context information.
   * @returns {Promise} A promise
   */

	}, {
		key: 'loginAsClientOwner',
		value: function loginAsClientOwner(_ref7) {
			var context = _ref7.context;

			return this.request({ uri: '/oauth/token', form: {
					grant_type: 'client_credentials',
					client_id: this.clientId,
					client_secret: this.clientSecret
				}, method: 'post', context: context });
		}

		/**
   * Create a user account for the Particle Cloud
   * @param  {Object} options Options for this API call
   * @param  {String} options.username Email of the new user
   * @param  {String} options.password Password
   * @param  {String} options.accountInfo Object that contains account information fields such as user real name, company name, business account flag etc
   * @returns {Promise} A promise
   */

	}, {
		key: 'createUser',
		value: function createUser(_ref8) {
			var username = _ref8.username,
			    password = _ref8.password,
			    accountInfo = _ref8.accountInfo,
			    context = _ref8.context;

			return this.post('/v1/users', {
				username: username,
				password: password,
				account_info: accountInfo
			}, undefined, context);
		}

		/**
   * Verify new user account via verification email
   * @param  {Object} options Options for this API call
   * @param  {String} options.token the string token sent in the verification email
   * @returns {Promise} A promise
   */

	}, {
		key: 'verifyUser',
		value: function verifyUser(_ref9) {
			var token = _ref9.token,
			    context = _ref9.context;

			return this.post('/v1/user/verify', {
				token: token
			}, undefined, context);
		}

		/**
   * Send reset password email for a Particle Cloud user account
   * @param  {Object} options Options for this API call
   * @param  {String} options.username Email of the user
   * @returns {Promise} A promise
   */

	}, {
		key: 'resetPassword',
		value: function resetPassword(_ref10) {
			var username = _ref10.username,
			    context = _ref10.context;

			return this.post('/v1/user/password-reset', { username: username }, undefined, context);
		}

		/**
   * Revoke an access token
   * @param  {Object} options Options for this API call
   * @param  {String} options.username Username of the Particle cloud account that the token belongs to.
   * @param  {String} options.password Password for the account
   * @param  {String} options.token    Access token you wish to revoke
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteAccessToken',
		value: function deleteAccessToken(_ref11) {
			var username = _ref11.username,
			    password = _ref11.password,
			    token = _ref11.token,
			    context = _ref11.context;

			return this.delete('/v1/access_tokens/' + token, {
				access_token: token
			}, { username: username, password: password }, context);
		}

		/**
   * Revoke the current session access token
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth         Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteCurrentAccessToken',
		value: function deleteCurrentAccessToken(_ref12) {
			var auth = _ref12.auth,
			    context = _ref12.context;

			return this.delete('/v1/access_tokens/current', undefined, auth, context);
		}

		/**
   * Delete the current user
   * @param {Object} options Options for this API call
   * @param {String} options.auth Access token
   * @param {String} options.password Password
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteUser',
		value: function deleteUser(_ref13) {
			var auth = _ref13.auth,
			    context = _ref13.context,
			    password = _ref13.password;

			return this.delete('/v1/user', { password: password }, auth, context);
		}

		/**
   * List all valid access tokens for a Particle Cloud account
   * @param  {Object} options Options for this API call
   * @param  {String} options.username Username
   * @param  {String} options.password Password
   * @returns {Promise} A promise
   */

	}, {
		key: 'listAccessTokens',
		value: function listAccessTokens(_ref14) {
			var username = _ref14.username,
			    password = _ref14.password,
			    context = _ref14.context;

			return this.get('/v1/access_tokens', { username: username, password: password }, undefined, context);
		}

		/**
   * Retrieves the information that is used to identify the current login for tracking.
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth      The access token
   * @param  {Boolean} options.full      When true, retrieve all information for registering a user with the tracking API. When false,
   *  retrieve only the unique tracking ID for the current login.
   * @param  {Object} options.context   Context information.
   * @returns {Promise<Object>} Resolve the tracking identify of the current login
   */

	}, {
		key: 'trackingIdentity',
		value: function trackingIdentity() {
			var _ref15 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
			    auth = _ref15.auth,
			    _ref15$full = _ref15.full,
			    full = _ref15$full === undefined ? false : _ref15$full,
			    context = _ref15.context;

			return this.get('/v1/user/identify', auth, full ? undefined : { tracking: 1 }, context);
		}

		/**
   * List devices claimed to the account or product
   * @param  {Object} options Options for this API call
   * @param  {String} [options.deviceId]   (Product only) Filter results to devices with this ID (partial matching)
   * @param  {String} [options.deviceName] (Product only) Filter results to devices with this name (partial matching)
   * @param  {Array.<string>} [options.groups]   (Product only) A list of full group names to filter results to devices belonging to these groups only.
   * @param  {String} [options.sortAttr]   (Product only) The attribute by which to sort results. See API docs for options.
   * @param  {String} [options.sortDir]    (Product only) The direction of sorting. See API docs for options.
   * @param  {Number} [options.page]       (Product only) Current page of results
   * @param  {Number} [options.perPage]    (Product only) Records per page
   * @param  {String} [options.product]    List devices in this product ID or slug
   * @param  {String} options.auth         Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listDevices',
		value: function listDevices(_ref16) {
			var deviceId = _ref16.deviceId,
			    deviceName = _ref16.deviceName,
			    groups = _ref16.groups,
			    sortAttr = _ref16.sortAttr,
			    sortDir = _ref16.sortDir,
			    page = _ref16.page,
			    perPage = _ref16.perPage,
			    product = _ref16.product,
			    auth = _ref16.auth,
			    context = _ref16.context;

			var uri = void 0,
			    query = void 0;

			if (product) {
				uri = '/v1/products/' + product + '/devices';
				groups = Array.isArray(groups) ? groups.join(',') : undefined;
				query = { deviceId: deviceId, deviceName: deviceName, groups: groups, sortAttr: sortAttr, sortDir: sortDir, page: page, per_page: perPage };
			} else {
				uri = '/v1/devices';
			}

			return this.get(uri, auth, query, context);
		}

		/**
   * Get detailed informationa about a device
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId  Device ID or Name
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth      Access token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getDevice',
		value: function getDevice(_ref17) {
			var deviceId = _ref17.deviceId,
			    product = _ref17.product,
			    auth = _ref17.auth,
			    context = _ref17.context;

			var uri = this.deviceUri({ deviceId: deviceId, product: product });
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Claim a device to the account. The device must be online and unclaimed.
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'claimDevice',
		value: function claimDevice(_ref18) {
			var deviceId = _ref18.deviceId,
			    requestTransfer = _ref18.requestTransfer,
			    auth = _ref18.auth,
			    context = _ref18.context;

			return this.post('/v1/devices', {
				id: deviceId,
				request_transfer: !!requestTransfer
			}, auth, context);
		}

		/**
   * Add a device to a product or move device out of quarantine.
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID
   * @param  {Object} options.file    A file that contains a single-column list of device IDs, device serial numbers, device IMEIs, or devie ICCIDs.
   *                                  Node: Either a path or Buffer. Browser: a File or Blob.
   * @param  {String} options.product  Add to this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'addDeviceToProduct',
		value: function addDeviceToProduct(_ref19) {
			var deviceId = _ref19.deviceId,
			    product = _ref19.product,
			    file = _ref19.file,
			    auth = _ref19.auth,
			    context = _ref19.context;

			var files = void 0,
			    data = void 0;

			if (file) {
				files = { file: file };
			} else if (deviceId) {
				data = { id: deviceId };
			}

			return this.request({
				uri: '/v1/products/' + product + '/devices',
				method: 'post',
				data: data,
				files: files,
				auth: auth,
				context: context
			});
		}

		/**
   * Unclaim / Remove a device from your account or product, or deny quarantine
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {Boolean} [options.deny]  (Product only) Deny this quarantined device, instead of removing an already approved device
   * @param  {String} options.product  Remove from this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'removeDevice',
		value: function removeDevice(_ref20) {
			var deviceId = _ref20.deviceId,
			    deny = _ref20.deny,
			    product = _ref20.product,
			    auth = _ref20.auth,
			    context = _ref20.context;

			var uri = this.deviceUri({ deviceId: deviceId, product: product });
			var data = product ? { deny: deny } : undefined;
			return this.delete(uri, data, auth, context);
		}

		/**
   * Unclaim a product device its the owner, but keep it in the product
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {String} options.product  Remove from this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'removeDeviceOwner',
		value: function removeDeviceOwner(_ref21) {
			var deviceId = _ref21.deviceId,
			    product = _ref21.product,
			    auth = _ref21.auth,
			    context = _ref21.context;

			var uri = '/v1/products/' + product + '/devices/' + deviceId + '/owner';
			return this.delete(uri, undefined, auth, context);
		}

		/**
   * Rename a device
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {String} options.name     Desired Name
   * @param  {String} [options.product] Rename device in this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'renameDevice',
		value: function renameDevice(_ref22) {
			var deviceId = _ref22.deviceId,
			    name = _ref22.name,
			    product = _ref22.product,
			    auth = _ref22.auth,
			    context = _ref22.context;

			return this.updateDevice({ deviceId: deviceId, name: name, product: product, auth: auth, context: context });
		}

		/**
   * Instruct the device to turn on/off the LED in a rainbow pattern
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {Boolean} options.signal   Signal on or off
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'signalDevice',
		value: function signalDevice(_ref23) {
			var deviceId = _ref23.deviceId,
			    signal = _ref23.signal,
			    product = _ref23.product,
			    auth = _ref23.auth,
			    context = _ref23.context;

			return this.updateDevice({ deviceId: deviceId, signal: signal, product: product, auth: auth, context: context });
		}

		/**
   * Store some notes about device
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId  Device ID or Name
   * @params {String} options.notes     Your notes about this device
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth      Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'setDeviceNotes',
		value: function setDeviceNotes(_ref24) {
			var deviceId = _ref24.deviceId,
			    notes = _ref24.notes,
			    product = _ref24.product,
			    auth = _ref24.auth,
			    context = _ref24.context;

			return this.updateDevice({ deviceId: deviceId, notes: notes, product: product, auth: auth, context: context });
		}

		/**
   * Mark device as being used in development of a product so it opts out of automatic firmware updates
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId      Device ID or Name
   * @param  {Boolean} options.development  Set to true to mark as development, false to return to product fleet
   * @param  {String} options.product       Device in this product ID or slug
   * @param  {String} options.auth          Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'markAsDevelopmentDevice',
		value: function markAsDevelopmentDevice(_ref25) {
			var deviceId = _ref25.deviceId,
			    _ref25$development = _ref25.development,
			    development = _ref25$development === undefined ? true : _ref25$development,
			    product = _ref25.product,
			    auth = _ref25.auth,
			    context = _ref25.context;

			return this.updateDevice({ deviceId: deviceId, development: development, product: product, auth: auth, context: context });
		}

		/**
   * Mark device as being used in development of a product so it opts out of automatic firmware updates
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId      Device ID or Name
   * @params {Number} options.desiredFirmwareVersion Lock the product device to run this firmware version.
   * @params {Boolean} [options.flash]      Immediately flash firmware indicated by desiredFirmwareVersion
   * @param  {String} options.product       Device in this product ID or slug
   * @param  {String} options.auth          Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'lockDeviceProductFirmware',
		value: function lockDeviceProductFirmware(_ref26) {
			var deviceId = _ref26.deviceId,
			    desiredFirmwareVersion = _ref26.desiredFirmwareVersion,
			    flash = _ref26.flash,
			    product = _ref26.product,
			    auth = _ref26.auth,
			    context = _ref26.context;

			return this.updateDevice({ deviceId: deviceId, desiredFirmwareVersion: desiredFirmwareVersion, flash: flash, product: product, auth: auth, context: context });
		}

		/**
   * Mark device as receiving automatic firmware updates
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId      Device ID or Name
   * @param  {String} options.product       Device in this product ID or slug
   * @param  {String} options.auth          Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'unlockDeviceProductFirmware',
		value: function unlockDeviceProductFirmware(_ref27) {
			var deviceId = _ref27.deviceId,
			    product = _ref27.product,
			    auth = _ref27.auth,
			    context = _ref27.context;

			return this.updateDevice({ deviceId: deviceId, desiredFirmwareVersion: null, product: product, auth: auth, context: context });
		}

		/**
   * Update multiple device attributes at the same time
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId       Device ID or Name
   * @param  {String} [options.name]         Desired Name
   * @param  {Boolean} options.signal        Signal device on or off
   * @params {String} [options.notes]        Your notes about this device
   * @param  {Boolean} [options.development] (Product only) Set to true to mark as development, false to return to product fleet
   * @params {Number} [options.desiredFirmwareVersion] (Product only) Lock the product device to run this firmware version.
   *                                              Pass `null` to unlock firmware and go back to released firmware.
   * @params {Boolean} [options.flash]       (Product only) Immediately flash firmware indicated by desiredFirmwareVersion
   * @param  {String} [options.product]      Device in this product ID or slug
   * @param  {String} options.auth           Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'updateDevice',
		value: function updateDevice(_ref28) {
			var deviceId = _ref28.deviceId,
			    name = _ref28.name,
			    signal = _ref28.signal,
			    notes = _ref28.notes,
			    development = _ref28.development,
			    desiredFirmwareVersion = _ref28.desiredFirmwareVersion,
			    flash = _ref28.flash,
			    product = _ref28.product,
			    auth = _ref28.auth,
			    context = _ref28.context;

			if (signal !== undefined) {
				signal = signal ? '1' : '0';
			}

			var uri = this.deviceUri({ deviceId: deviceId, product: product });
			var data = product ? { name: name, signal: signal, notes: notes, development: development, desired_firmware_version: desiredFirmwareVersion, flash: flash } : { name: name, signal: signal, notes: notes };
			return this.put(uri, data, auth, context);
		}

		/**
   * Provision a new device for products that allow self-provisioning
   * @param  {Object} options Options for this API call
   * @param  {String} options.productId Product ID where to create this device
   * @param  {String} options.auth      Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'provisionDevice',
		value: function provisionDevice(_ref29) {
			var productId = _ref29.productId,
			    auth = _ref29.auth,
			    context = _ref29.context;

			return this.post('/v1/devices', { product_id: productId }, auth, context);
		}

		/**
   * Generate a claim code to use in the device claiming process.
   * To generate a claim code for a product, the access token MUST belong to a
   * customer of the product.
   * @param  {Object} options Options for this API call
   * @param  {String} [options.iccid] ICCID of the SIM card used in the Electron
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth  Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getClaimCode',
		value: function getClaimCode(_ref30) {
			var iccid = _ref30.iccid,
			    product = _ref30.product,
			    auth = _ref30.auth,
			    context = _ref30.context;

			var uri = product ? '/v1/products/' + product + '/device_claims' : '/v1/device_claims';
			return this.post(uri, { iccid: iccid }, auth, context);
		}
	}, {
		key: 'validatePromoCode',
		value: function validatePromoCode(_ref31) {
			var auth = _ref31.auth,
			    promoCode = _ref31.promoCode,
			    context = _ref31.context;

			return this.get('/v1/promo_code/' + promoCode, auth, undefined, context);
		}
	}, {
		key: 'changeProduct',
		value: function changeProduct(_ref32) {
			var deviceId = _ref32.deviceId,
			    productId = _ref32.productId,
			    auth = _ref32.auth,
			    context = _ref32.context;

			return this.put('/v1/devices/' + deviceId, {
				product_id: productId
			}, auth, context);
		}

		/**
   * Get the value of a device variable
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {String} options.name     Variable name
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getVariable',
		value: function getVariable(_ref33) {
			var deviceId = _ref33.deviceId,
			    name = _ref33.name,
			    product = _ref33.product,
			    auth = _ref33.auth,
			    context = _ref33.context;

			var uri = product ? '/v1/products/' + product + '/devices/' + deviceId + '/' + name : '/v1/devices/' + deviceId + '/' + name;
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Compile and flash application firmware to a device. Pass a pre-compiled binary to flash it directly to the device.
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId      Device ID or Name
   * @param  {String} options.product       Flash device in this product ID or slug
   * @param  {Object} options.files         Object containing files to be compiled and flashed. Keys should be the filenames, including relative path, and the values should be a path or Buffer of the file contents in Node, or a File or Blob in the browser.
   * @param  {String} [options.targetVersion=latest] System firmware version to compile against
   * @param  {String} options.auth          String
   * @returns {Promise} A promise
   */

	}, {
		key: 'flashDevice',
		value: function flashDevice(_ref34) {
			var deviceId = _ref34.deviceId,
			    product = _ref34.product,
			    files = _ref34.files,
			    targetVersion = _ref34.targetVersion,
			    auth = _ref34.auth,
			    context = _ref34.context;

			var uri = this.deviceUri({ deviceId: deviceId, product: product });
			var form = {};
			if (targetVersion) {
				form.build_target_version = targetVersion;
			} else {
				form.latest = 'true';
			}
			return this.request({ uri: uri, files: files, auth: auth, form: form, context: context, method: 'put' });
		}

		/**
   * DEPRECATED: Flash the Tinker application to a device. Instead compile and flash the Tinker source code.
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'flashTinker',
		value: function flashTinker(_ref35) {
			var deviceId = _ref35.deviceId,
			    auth = _ref35.auth,
			    context = _ref35.context;

			/* eslint-disable no-console */
			if (console && console.warning) {
				console.warning('Particle.flashTinker is deprecated');
			}
			/* eslint-enable no-console */
			return this.put('/v1/devices/' + deviceId, {
				app: 'tinker'
			}, auth, context);
		}

		/**
   * Compile firmware using the Particle Cloud
   * @param  {Object} options Options for this API call
   * @param  {Object} options.files         Object containing files to be compiled. Keys should be the filenames, including relative path, and the values should be a path or Buffer of the file contents in Node, or a File or Blob in the browser.
   * @param  {Number} [options.platformId]    Platform id number of the device you are compiling for. Common values are 0=Core, 6=Photon, 10=Electron.
   * @param  {String} [options.targetVersion=latest] System firmware version to compile against
   * @param  {String} options.auth          Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'compileCode',
		value: function compileCode(_ref36) {
			var files = _ref36.files,
			    platformId = _ref36.platformId,
			    targetVersion = _ref36.targetVersion,
			    auth = _ref36.auth,
			    context = _ref36.context;

			var form = { platform_id: platformId };
			if (targetVersion) {
				form.build_target_version = targetVersion;
			} else {
				form.latest = 'true';
			}
			return this.request({
				uri: '/v1/binaries',
				method: 'post',
				files: files,
				auth: auth,
				form: form,
				context: context
			});
		}

		/**
   * Download a firmware binary
   * @param  {Object} options Options for this API call
   * @param  {String} options.binaryId Binary ID received from a successful compile call
   * @param  {String} options.auth     Access Token
   * @returns {Request} A promise
   */

	}, {
		key: 'downloadFirmwareBinary',
		value: function downloadFirmwareBinary(_ref37) {
			var binaryId = _ref37.binaryId,
			    auth = _ref37.auth;

			var req = this.request({
				uri: '/v1/binaries/' + binaryId,
				method: 'get',
				raw: true,
				auth: auth
			});

			return this._provideFileData(req);
		}

		/**
   * Send a new device public key to the Particle Cloud
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId  Device ID or Name
   * @param  {(String|Buffer)} options.key       Public key contents
   * @param  {String} [options.algorithm=rsa] Algorithm used to generate the public key. Valid values are `rsa` or `ecc`.
   * @param  {String} options.auth      Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'sendPublicKey',
		value: function sendPublicKey(_ref38) {
			var deviceId = _ref38.deviceId,
			    key = _ref38.key,
			    algorithm = _ref38.algorithm,
			    auth = _ref38.auth,
			    context = _ref38.context;

			return this.post('/v1/provisioning/' + deviceId, {
				deviceID: deviceId,
				publicKey: typeof key === 'string' ? key : key.toString(),
				filename: 'particle-api',
				order: 'manual_' + Date.now(),
				algorithm: algorithm || 'rsa'
			}, auth, context);
		}

		/**
   * Call a device function
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId Device ID or Name
   * @param  {String} options.name     Function name
   * @param  {String} options.argument Function argument
   * @param  {String} [options.product] Device in this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'callFunction',
		value: function callFunction(_ref39) {
			var deviceId = _ref39.deviceId,
			    name = _ref39.name,
			    argument = _ref39.argument,
			    product = _ref39.product,
			    auth = _ref39.auth,
			    context = _ref39.context;

			var uri = product ? '/v1/products/' + product + '/devices/' + deviceId + '/' + name : '/v1/devices/' + deviceId + '/' + name;
			return this.post(uri, { args: argument }, auth, context);
		}

		/**
   * Get a stream of events
   * @param  {Object} options Options for this API call
   * @param  {String} [options.deviceId] Device ID or Name, or `mine` to indicate only your devices.
   * @param  {String} [options.name]     Event Name
   * @param  {String} [options.org]     Organization Slug
   * @param  {String} [options.product] Events for this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} If the promise resolves, the resolution value will be an EventStream object that will
   * emit 'event' events.
   */

	}, {
		key: 'getEventStream',
		value: function getEventStream(_ref40) {
			var deviceId = _ref40.deviceId,
			    name = _ref40.name,
			    org = _ref40.org,
			    product = _ref40.product,
			    auth = _ref40.auth;

			var uri = '/v1/';
			if (org) {
				uri += 'orgs/' + org + '/';
			}

			if (product) {
				uri += 'products/' + product + '/';
			}

			if (deviceId) {
				uri += 'devices/';
				if (!(deviceId.toLowerCase() === 'mine')) {
					uri += deviceId + '/';
				}
			}

			uri += 'events';

			if (name) {
				uri += '/' + encodeURIComponent(name);
			}

			return new _EventStream2.default('' + this.baseUrl + uri, auth).connect();
		}

		/**
   * Publish a event to the Particle Cloud
   * @param  {Object} options Options for this API call
   * @param  {String} options.name      Event name
   * @param  {String} options.data      Event data
   * @param  {Boolean} options.isPrivate Should the event be publicly available?
   * @param  {String} [options.product]  Event for this product ID or slug
   * @param  {String} options.auth      Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'publishEvent',
		value: function publishEvent(_ref41) {
			var name = _ref41.name,
			    data = _ref41.data,
			    isPrivate = _ref41.isPrivate,
			    product = _ref41.product,
			    auth = _ref41.auth,
			    context = _ref41.context;

			var uri = product ? '/v1/products/' + product + '/events' : '/v1/devices/events';
			var postData = { name: name, data: data, private: isPrivate };
			return this.post(uri, postData, auth, context);
		}

		/**
   * Create a webhook
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId           Trigger webhook only for this device ID or Name
   * @param  {String} options.name               Webhook name
   * @param  {String} options.url                URL the webhook should hit
   * @param  {String} [options.requestType=POST]        HTTP method to use
   * @param  {Object} [options.headers]            Additional headers to add to the webhook
   * @param  {Object} [options.json]               JSON data
   * @param  {Object} [options.query]              Query string data
   * @param  {String} [options.body]               Custom webhook request body
   * @param  {Object} [options.responseTemplate]   Webhook response template
   * @param  {Object} [options.responseTopic]      Webhook response topic
   * @param  {Boolean} [options.rejectUnauthorized] Reject invalid HTTPS certificates
   * @params {Boolean} [options.noDefaults]        Don't include default event data in the webhook request
   * @param  {Object} [options.webhookAuth]        HTTP Basic Auth information
   * @param  {Object} [options.form]               Form data
   * @param  {String} [options.product]          Webhook for this product ID or slug
   * @param  {String} options.auth               Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'createWebhook',
		value: function createWebhook(_ref42) {
			var deviceId = _ref42.deviceId,
			    name = _ref42.name,
			    url = _ref42.url,
			    requestType = _ref42.requestType,
			    headers = _ref42.headers,
			    json = _ref42.json,
			    query = _ref42.query,
			    body = _ref42.body,
			    responseTemplate = _ref42.responseTemplate,
			    responseTopic = _ref42.responseTopic,
			    rejectUnauthorized = _ref42.rejectUnauthorized,
			    webhookAuth = _ref42.webhookAuth,
			    noDefaults = _ref42.noDefaults,
			    form = _ref42.form,
			    product = _ref42.product,
			    auth = _ref42.auth,
			    context = _ref42.context;

			// deviceId: 'mine' is deprecated since webhooks only trigger on your device anyways
			if (deviceId === 'mine') {
				deviceId = undefined;
			}
			var uri = product ? '/v1/products/' + product + '/webhooks' : '/v1/webhooks';
			var data = { event: name, deviceid: deviceId, url: url, requestType: requestType, headers: headers, json: json, query: query, body: body, responseTemplate: responseTemplate, responseTopic: responseTopic, rejectUnauthorized: rejectUnauthorized, auth: webhookAuth, noDefaults: noDefaults, form: form };
			return this.post(uri, data, auth, context);
		}

		/**
   * Delete a webhook
   * @param  {Object} options Options for this API call
   * @param  {String} options.hookId Webhook ID
   * @param  {String} [options.product]          Webhook for this product ID or slug
   * @param  {String} options.auth   Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteWebhook',
		value: function deleteWebhook(_ref43) {
			var hookId = _ref43.hookId,
			    product = _ref43.product,
			    auth = _ref43.auth,
			    context = _ref43.context;

			var uri = product ? '/v1/products/' + product + '/webhooks/' + hookId : '/v1/webhooks/' + hookId;
			return this.delete(uri, undefined, auth, context);
		}

		/**
   * List all webhooks owned by the account or product
   * @param  {Object} options Options for this API call
   * @param  {String} [options.product]          Webhooks for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listWebhooks',
		value: function listWebhooks(_ref44) {
			var product = _ref44.product,
			    auth = _ref44.auth,
			    context = _ref44.context;

			var uri = product ? '/v1/products/' + product + '/webhooks' : '/v1/webhooks';
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Create an integration to send events to an external service
      *
   * See the API docs for details https://docs.particle.io/reference/api/#integrations-webhooks-
   *
   * @param  {Object} options Options for this API call
   * @param  {String} options.event            Event that triggers the integration
   * @params {Object} options.settings         Settings specific to that integration type
   * @param  {String} [options.deviceId]       Trigger integration only for this device ID or Name
   * @param  {String} [options.product]        Integration for this product ID or slug
   * @param  {String} options.auth             Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'createIntegration',
		value: function createIntegration(_ref45) {
			var event = _ref45.event,
			    settings = _ref45.settings,
			    deviceId = _ref45.deviceId,
			    product = _ref45.product,
			    auth = _ref45.auth,
			    context = _ref45.context;

			var uri = product ? '/v1/products/' + product + '/integrations' : '/v1/integrations';
			var data = (0, _assign2.default)({ event: event, deviceid: deviceId }, settings);
			return this.post(uri, data, auth, context);
		}

		/**
   * Edit an integration to send events to an external service
   *
   * See the API docs for details https://docs.particle.io/reference/api/#integrations-webhooks-
   *
   * @param  {Object} options Options for this API call
   * @param  {String} options.integrationId    The integration to edit
   * @param  {String} [options.event]          Change the event that triggers the integration
   * @params {Object} [options.settings]       Change the settings specific to that integration type
   * @param  {String} [options.deviceId]       Trigger integration only for this device ID or Name
   * @param  {String} [options.product]        Integration for this product ID or slug
   * @param  {String} options.auth             Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'editIntegration',
		value: function editIntegration(_ref46) {
			var integrationId = _ref46.integrationId,
			    event = _ref46.event,
			    settings = _ref46.settings,
			    deviceId = _ref46.deviceId,
			    product = _ref46.product,
			    auth = _ref46.auth,
			    context = _ref46.context;

			var uri = product ? '/v1/products/' + product + '/integrations/' + integrationId : '/v1/integrations/' + integrationId;
			var data = (0, _assign2.default)({ event: event, deviceid: deviceId }, settings);
			return this.put(uri, data, auth, context);
		}

		/**
   * Delete an integration to send events to an external service
   *
   * @param  {Object} options Options for this API call
   * @param  {String} options.integrationId    The integration to remove
   * @param  {String} [options.product]        Integration for this product ID or slug
   * @param  {String} options.auth             Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteIntegration',
		value: function deleteIntegration(_ref47) {
			var integrationId = _ref47.integrationId,
			    product = _ref47.product,
			    auth = _ref47.auth,
			    context = _ref47.context;

			var uri = product ? '/v1/products/' + product + '/integrations/' + integrationId : '/v1/integrations/' + integrationId;
			return this.delete(uri, undefined, auth, context);
		}

		/**
   * List all integrations owned by the account or product
   * @param  {Object} options Options for this API call
   * @param  {String} [options.product]        Integrations for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listIntegrations',
		value: function listIntegrations(_ref48) {
			var product = _ref48.product,
			    auth = _ref48.auth,
			    context = _ref48.context;

			var uri = product ? '/v1/products/' + product + '/integrations' : '/v1/integrations';
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Get details about the current user
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getUserInfo',
		value: function getUserInfo(_ref49) {
			var auth = _ref49.auth,
			    context = _ref49.context;

			return this.get('/v1/user', auth, undefined, context);
		}

		/**
   * Set details on the current user
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth Access Token
   * @param  {String} options.accountInfo Set user's extended info fields (name, business account, company name, etc)
   * @returns {Promise} A promise
   */

	}, {
		key: 'setUserInfo',
		value: function setUserInfo(_ref50) {
			var accountInfo = _ref50.accountInfo,
			    auth = _ref50.auth,
			    context = _ref50.context;

			var bodyObj = {
				account_info: accountInfo
			};

			return this.put('/v1/user', bodyObj, auth, context);
		}

		/**
   * Change username (i.e, email)
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth Access Token
   * @param  {String} options.currentPassword Current password
   * @param  {String} options.username New email
   * @returns {Promise} A promise
   */

	}, {
		key: 'changeUsername',
		value: function changeUsername(_ref51) {
			var currentPassword = _ref51.currentPassword,
			    username = _ref51.username,
			    auth = _ref51.auth,
			    context = _ref51.context;

			var bodyObj = {
				current_password: currentPassword,
				username: username
			};

			return this.put('/v1/user', bodyObj, auth, context);
		}

		/**
   * Change user's password
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth Access Token
   * @param  {String} options.currentPassword Current password
   * @param  {String} options.password New password
   * @param  {Boolean} options.invalidateTokens Should all tokens be invalidated
   * @returns {Promise} A promise
   */

	}, {
		key: 'changeUserPassword',
		value: function changeUserPassword(_ref52) {
			var currentPassword = _ref52.currentPassword,
			    password = _ref52.password,
			    _ref52$invalidateToke = _ref52.invalidateTokens,
			    invalidateTokens = _ref52$invalidateToke === undefined ? false : _ref52$invalidateToke,
			    auth = _ref52.auth,
			    context = _ref52.context;

			var bodyObj = {
				current_password: currentPassword,
				password: password
			};

			if (invalidateTokens) {
				bodyObj.invalidate_tokens = true;
			}

			return this.put('/v1/user', bodyObj, auth, context);
		}

		/**
   * List SIM cards owned by a user or product
   * @param  {Object} options Options for this API call
   * @param  {String} [options.iccid]    (Product only) Filter to SIM cards matching this ICCID
   * @param  {String} [options.deviceId] (Product only) Filter to SIM cards matching this device ID
   * @param  {String} [options.deviceName] (Product only) Filter to SIM cards matching this device name
   * @param  {Number} [options.page]     (Product only) Current page of results
   * @param  {Number} [options.perPage]  (Product only) Records per page
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listSIMs',
		value: function listSIMs(_ref53) {
			var iccid = _ref53.iccid,
			    deviceId = _ref53.deviceId,
			    deviceName = _ref53.deviceName,
			    page = _ref53.page,
			    perPage = _ref53.perPage,
			    product = _ref53.product,
			    auth = _ref53.auth,
			    context = _ref53.context;

			var uri = product ? '/v1/products/' + product + '/sims' : '/v1/sims';
			var query = product ? { iccid: iccid, deviceId: deviceId, deviceName: deviceName, page: page, per_page: perPage } : undefined;
			return this.get(uri, auth, query, context);
		}

		/**
   * Get data usage for one SIM card for the current billing period
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid      ICCID of the SIM card
   * @param  {String} [options.product]  SIM card for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getSIMDataUsage',
		value: function getSIMDataUsage(_ref54) {
			var iccid = _ref54.iccid,
			    product = _ref54.product,
			    auth = _ref54.auth,
			    context = _ref54.context;

			var uri = product ? '/v1/products/' + product + '/sims/' + iccid + '/data_usage' : '/v1/sims/' + iccid + '/data_usage';
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Get data usage for all SIM cards in a product the current billing period
   * @param  {Object} options Options for this API call
   * @param  {String} options.product  SIM cards for this product ID or slug
   * @param  {String} options.auth     Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getFleetDataUsage',
		value: function getFleetDataUsage(_ref55) {
			var product = _ref55.product,
			    auth = _ref55.auth,
			    context = _ref55.context;

			return this.get('/v1/products/' + product + '/sims/data_usage', auth, undefined, context);
		}
	}, {
		key: 'checkSIM',
		value: function checkSIM(_ref56) {
			var iccid = _ref56.iccid,
			    auth = _ref56.auth,
			    context = _ref56.context;

			return this.head('/v1/sims/' + iccid, auth, undefined, context);
		}

		/**
   * Activate and add SIM cards to an account or product
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid        ICCID of the SIM card
   * @param  {Array<String>} options.iccids (Product only) ICCID of multiple SIM cards to import
   * @param  {String} options.country The ISO country code for the SIM cards
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'activateSIM',
		value: function activateSIM(_ref57) {
			var iccid = _ref57.iccid,
			    iccids = _ref57.iccids,
			    country = _ref57.country,
			    promoCode = _ref57.promoCode,
			    product = _ref57.product,
			    auth = _ref57.auth,
			    context = _ref57.context;

			// promoCode is deprecated
			iccids = iccids || [iccid];
			var uri = product ? '/v1/products/' + product + '/sims' : '/v1/sims/' + iccid;
			var data = product ? { sims: iccids, country: country } : { country: country, promoCode: promoCode, action: 'activate' };
			var method = product ? 'post' : 'put';

			return this.request({ uri: uri, method: method, data: data, auth: auth, context: context });
		}

		/**
   * Deactivate a SIM card so it doesn't incur data usage in future months.
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid      ICCID of the SIM card
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deactivateSIM',
		value: function deactivateSIM(_ref58) {
			var iccid = _ref58.iccid,
			    product = _ref58.product,
			    auth = _ref58.auth,
			    context = _ref58.context;

			var uri = product ? '/v1/products/' + product + '/sims/' + iccid : '/v1/sims/' + iccid;
			var data = { action: 'deactivate' };
			return this.put(uri, data, auth, context);
		}

		/**
   * Reactivate a SIM card the was deactivated or unpause a SIM card that was automatically paused
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid      ICCID of the SIM card
   * @param  {Number} [options.mbLimit]  New monthly data limit. Necessary if unpausing a SIM card
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'reactivateSIM',
		value: function reactivateSIM(_ref59) {
			var iccid = _ref59.iccid,
			    mbLimit = _ref59.mbLimit,
			    product = _ref59.product,
			    auth = _ref59.auth,
			    context = _ref59.context;

			var uri = product ? '/v1/products/' + product + '/sims/' + iccid : '/v1/sims/' + iccid;
			var data = { mb_limit: mbLimit, action: 'reactivate' };
			return this.put(uri, data, auth, context);
		}

		/**
   * Update SIM card data limit
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid        ICCID of the SIM card
   * @param  {Array}  options.mbLimit     Data limit in megabyte for the SIM card
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'updateSIM',
		value: function updateSIM(_ref60) {
			var iccid = _ref60.iccid,
			    mbLimit = _ref60.mbLimit,
			    product = _ref60.product,
			    auth = _ref60.auth,
			    context = _ref60.context;

			var uri = product ? '/v1/products/' + product + '/sims/' + iccid : '/v1/sims/' + iccid;
			var data = { mb_limit: mbLimit };
			return this.put(uri, data, auth, context);
		}

		/**
   * Remove a SIM card from an account so it can be activated by a different account
   * @param  {Object} options Options for this API call
   * @param  {String} options.iccid      ICCID of the SIM card
   * @param  {String} [options.product]  SIM cards for this product ID or slug
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'removeSIM',
		value: function removeSIM(_ref61) {
			var iccid = _ref61.iccid,
			    product = _ref61.product,
			    auth = _ref61.auth,
			    context = _ref61.context;

			var uri = product ? '/v1/products/' + product + '/sims/' + iccid : '/v1/sims/' + iccid;
			return this.delete(uri, undefined, auth, context);
		}

		/**
   * List valid build targets to be used for compiling
   * @param  {Object} options Options for this API call
   * @param  {Boolean} [options.onlyFeatured=false] Only list featured build targets
   * @param  {String} options.auth       Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listBuildTargets',
		value: function listBuildTargets(_ref62) {
			var onlyFeatured = _ref62.onlyFeatured,
			    auth = _ref62.auth,
			    context = _ref62.context;

			var query = void 0;
			if (onlyFeatured !== undefined) {
				query = { featured: !!onlyFeatured };
			}
			return this.get('/v1/build_targets', auth, query, context);
		}

		/**
   * List firmware libraries
   * @param  {Object} options Options for this API call
   * @param  {Number} options.page Page index (default, first page)
   * @param  {Number} options.limit Number of items per page
   * @param  {String} options.filter Search term for the libraries
   * @param  {String} options.sort Ordering key for the library list
   * @param  {Array<String>}  options.architectures List of architectures to filter
   * @param  {String} options.category Category to filter
   * @param  {String} options.scope The library scope to list. Default is 'all'. Other values are
   * - 'all' - list public libraries and my private libraries
   * - 'public' - list only public libraries
   * - 'private' - list only my private libraries
   * - 'mine' - list my libraries (public and private)
   * - 'official' - list only official libraries
   * - 'verified' - list only verified libraries
   * - 'featured' - list only featured libraries
   * @param  {String} options.excludeScopes  list of scopes to exclude
   * @param  {String} options.category Category to filter
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listLibraries',
		value: function listLibraries(_ref63) {
			var page = _ref63.page,
			    limit = _ref63.limit,
			    filter = _ref63.filter,
			    sort = _ref63.sort,
			    architectures = _ref63.architectures,
			    category = _ref63.category,
			    scope = _ref63.scope,
			    excludeScopes = _ref63.excludeScopes,
			    auth = _ref63.auth,
			    context = _ref63.context;

			return this.get('/v1/libraries', auth, {
				page: page,
				filter: filter,
				limit: limit,
				sort: sort,
				architectures: this._asList(architectures),
				category: category,
				scope: scope,
				excludeScopes: this._asList(excludeScopes)
			}, context);
		}
	}, {
		key: '_asList',
		value: function _asList(value) {
			return Array.isArray(value) ? value.join(',') : value;
		}

		/**
   * Get firmware library details
   * @param  {Object} options Options for this API call
   * @param  {String} options.name Name of the library to fetch
   * @param  {String} options.version Version of the library to fetch (default: latest)
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getLibrary',
		value: function getLibrary(_ref64) {
			var name = _ref64.name,
			    version = _ref64.version,
			    auth = _ref64.auth,
			    context = _ref64.context;

			return this.get('/v1/libraries/' + name, auth, { version: version }, context);
		}

		/**
   * Firmware library details for each version
   * @param  {Object} options Options for this API call
   * @param  {String} options.name Name of the library to fetch
   * @param  {Number} options.page Page index (default, first page)
   * @param  {Number} options.limit Number of items per page
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getLibraryVersions',
		value: function getLibraryVersions(_ref65) {
			var name = _ref65.name,
			    page = _ref65.page,
			    limit = _ref65.limit,
			    auth = _ref65.auth,
			    context = _ref65.context;

			return this.get('/v1/libraries/' + name + '/versions', auth, {
				page: page,
				limit: limit
			}, context);
		}

		/**
   * Contribute a new library version from a compressed archive
   * @param  {Object} options Options for this API call
   * @param  {String} options.archive Compressed archive file containing the library sources
   *                                  Either a path or Buffer of the file contents in Node, or a File or Blob in the browser.
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'contributeLibrary',
		value: function contributeLibrary(_ref66) {
			var archive = _ref66.archive,
			    auth = _ref66.auth,
			    context = _ref66.context;

			var files = {
				'archive.tar.gz': archive
			};

			return this.request({
				uri: '/v1/libraries',
				method: 'post',
				files: files,
				auth: auth,
				context: context
			});
		}

		/**
   * Publish the latest version of a library to the public
   * @param  {Object} options Options for this API call
   * @param  {String} options.name Name of the library to publish
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'publishLibrary',
		value: function publishLibrary(_ref67) {
			var name = _ref67.name,
			    auth = _ref67.auth,
			    context = _ref67.context;

			return this.request({
				uri: '/v1/libraries/' + name,
				method: 'patch',
				data: { visibility: 'public' },
				auth: auth,
				context: context
			});
		}

		/**
   * Delete one version of a library or an entire private library
   * @param  {Object} options Options for this API call
   * @param  {String} options.name Name of the library to remove
   * @param  {String} options.force Key to force deleting a public library
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteLibrary',
		value: function deleteLibrary(_ref68) {
			var name = _ref68.name,
			    force = _ref68.force,
			    auth = _ref68.auth,
			    context = _ref68.context;

			return this.delete('/v1/libraries/' + name, { force: force }, auth, context);
		}

		/**
   * Download an external file that may not be on the API
   * @param  {Object} options Options for this API call
   * @param  {String} options.url URL of the file.
   * @returns {Promise} Resolves to a buffer with the file data
   */

	}, {
		key: 'downloadFile',
		value: function downloadFile(_ref69) {
			var url = _ref69.url;

			var req = this.request({ uri: url, method: 'get', raw: true });
			return this._provideFileData(req);
		}

		/**
   * List OAuth client created by the account
   * @param  {Object} options Options for this API call
   * @param  {String} [options.product] List clients for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listOAuthClients',
		value: function listOAuthClients(_ref70) {
			var product = _ref70.product,
			    auth = _ref70.auth,
			    context = _ref70.context;

			var uri = product ? '/v1/products/' + product + '/clients' : '/v1/clients';
			return this.get(uri, auth, undefined, context);
		}

		/**
   * Create an OAuth client
   * @param  {Object} options Options for this API call
   * @param  {String} options.name               Name of the OAuth client
   * @param  {String} options.type               web, installed or web
   * @param  {String} [options.redirect_uri]     URL to redirect after OAuth flow. Only for type web.
   * @param  {Object} [options.scope]            Limits what the access tokens created by this client can do.
   * @param  {String} [options.product]          Create client for this product ID or slug
   * @param  {String} options.auth               Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'createOAuthClient',
		value: function createOAuthClient(_ref71) {
			var name = _ref71.name,
			    type = _ref71.type,
			    redirect_uri = _ref71.redirect_uri,
			    scope = _ref71.scope,
			    product = _ref71.product,
			    auth = _ref71.auth,
			    context = _ref71.context;

			var uri = product ? '/v1/products/' + product + '/clients' : '/v1/clients';
			var data = { name: name, type: type, redirect_uri: redirect_uri, scope: scope };
			return this.post(uri, data, auth, context);
		}

		/**
   * Update an OAuth client
   * @param  {Object} options Options for this API call
   * @param  {String} options.clientId           The OAuth client to update
   * @param  {String} [options.name]             New Name of the OAuth client
   * @param  {Object} [options.scope]            New scope of the OAuth client
   * @param  {String} [options.product]          Update client linked to this product ID or slug
   * @param  {String} options.auth               Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'updateOAuthClient',
		value: function updateOAuthClient(_ref72) {
			var clientId = _ref72.clientId,
			    name = _ref72.name,
			    scope = _ref72.scope,
			    product = _ref72.product,
			    auth = _ref72.auth,
			    context = _ref72.context;

			var uri = product ? '/v1/products/' + product + '/clients/' + clientId : '/v1/clients/' + clientId;
			var data = { name: name, scope: scope };
			return this.put(uri, data, auth, context);
		}

		/**
   * Delete an OAuth client
   * @param  {Object} options Options for this API call
   * @param  {String} options.clientId           The OAuth client to update
   * @param  {String} [options.product]          OAuth client linked to this product ID or slug
   * @param  {String} options.auth               Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'deleteOAuthClient',
		value: function deleteOAuthClient(_ref73) {
			var clientId = _ref73.clientId,
			    product = _ref73.product,
			    auth = _ref73.auth,
			    context = _ref73.context;

			var uri = product ? '/v1/products/' + product + '/clients/' + clientId : '/v1/clients/' + clientId;
			return this.delete(uri, undefined, auth, context);
		}

		/**
   * List products the account has access to
   * @param  {Object} options Options for this API call
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listProducts',
		value: function listProducts(_ref74) {
			var auth = _ref74.auth,
			    context = _ref74.context;

			return this.get('/v1/products', auth, undefined, context);
		}

		/**
   * Get detailed information about a product
   * @param  {Object} options Options for this API call
   * @param  {String} options.product  Product ID or slug
   * @param  {String} options.auth     Access token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getProduct',
		value: function getProduct(_ref75) {
			var product = _ref75.product,
			    auth = _ref75.auth,
			    context = _ref75.context;

			return this.get('/v1/products/' + product, auth, undefined, context);
		}

		/**
   * List product firmware versions
   * @param  {Object} options Options for this API call
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listProductFirmware',
		value: function listProductFirmware(_ref76) {
			var product = _ref76.product,
			    auth = _ref76.auth,
			    context = _ref76.context;

			return this.get('/v1/products/' + product + '/firmware', auth, undefined, context);
		}

		/**
   * List product firmware versions
   * @param  {Object} options Options for this API call
   * @param  {Object} options.file    Path or Buffer of the new firmware file
   *                                  Either a path or Buffer of the file contents in Node, or a File or Blob in the browser.
   * @param  {Number} options.version Version number of new firmware
   * @param  {String} options.title   Short identifier for the new firmware
   * @param  {String} [options.description] Longer description for the new firmware
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'uploadProductFirmware',
		value: function uploadProductFirmware(_ref77) {
			var file = _ref77.file,
			    version = _ref77.version,
			    title = _ref77.title,
			    description = _ref77.description,
			    product = _ref77.product,
			    auth = _ref77.auth,
			    context = _ref77.context;

			return this.request({
				uri: '/v1/products/' + product + '/firmware',
				method: 'post',
				files: {
					'firmware.bin': file
				},
				form: {
					version: version,
					title: title,
					description: description
				},
				context: context,
				auth: auth
			});
		}

		/**
   * Get information about a product firmware version
   * @param  {Object} options Options for this API call
   * @param  {Number} options.version Version number of firmware
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth    Access token
   * @returns {Promise} A promise
   */

	}, {
		key: 'getProductFirmware',
		value: function getProductFirmware(_ref78) {
			var version = _ref78.version,
			    product = _ref78.product,
			    auth = _ref78.auth,
			    context = _ref78.context;

			return this.get('/v1/products/' + product + '/firmware/' + version, auth, undefined, context);
		}

		/**
   * Update information for a product firmware version
   * @param  {Object} options Options for this API call
   * @param  {Number} options.version Version number of new firmware
   * @param  {String} [options.title]   New title
   * @param  {String} [options.description] New description
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'updateProductFirmware',
		value: function updateProductFirmware(_ref79) {
			var version = _ref79.version,
			    title = _ref79.title,
			    description = _ref79.description,
			    product = _ref79.product,
			    auth = _ref79.auth,
			    context = _ref79.context;

			var uri = '/v1/products/' + product + '/firmware/' + version;
			return this.put(uri, { title: title, description: description }, auth, context);
		}

		/**
   * Download a product firmware binary
   * @param  {Object} options Options for this API call
   * @param  {Number} options.version Version number of new firmware
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth    Access Token
   * @returns {Request} A promise
   */

	}, {
		key: 'downloadProductFirmware',
		value: function downloadProductFirmware(_ref80) {
			var version = _ref80.version,
			    product = _ref80.product,
			    auth = _ref80.auth;

			var req = this.request({
				uri: '/v1/products/' + product + '/firmware/' + version + '/binary',
				method: 'get',
				raw: true,
				auth: auth
			});

			return this._provideFileData(req);
		}
	}, {
		key: '_provideFileData',
		value: function _provideFileData(req) {
			if (this.agent.isForBrowser()) {
				req = req.responseType('arraybuffer').then(function (res) {
					res.body = res.xhr.response;
					return res;
				});
			} else {
				req = req.buffer(true).parse(_superagentBinaryParser2.default);
			}
			return req.then(function (res) {
				return res.body;
			});
		}

		/**
   * Release a product firmware version as the default version
   * @param  {Object} options Options for this API call
   * @param  {Number} options.version Version number of new firmware
   * @param  {String} options.product Firmware for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'releaseProductFirmware',
		value: function releaseProductFirmware(_ref81) {
			var version = _ref81.version,
			    product = _ref81.product,
			    auth = _ref81.auth,
			    context = _ref81.context;

			var uri = '/v1/products/' + product + '/firmware/release';
			return this.put(uri, { version: version }, auth, context);
		}

		/**
   * List product team members
   * @param  {Object} options Options for this API call
   * @param  {String} options.product Team for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'listTeamMembers',
		value: function listTeamMembers(_ref82) {
			var product = _ref82.product,
			    auth = _ref82.auth,
			    context = _ref82.context;

			return this.get('/v1/products/' + product + '/team', auth, undefined, context);
		}

		/**
   * Invite Particle user to a product team
   * @param  {Object} options Options for this API call
   * @param  {String} options.username  Username for the Particle account
   * @param  {String} options.product Team for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'inviteTeamMember',
		value: function inviteTeamMember(_ref83) {
			var username = _ref83.username,
			    product = _ref83.product,
			    auth = _ref83.auth,
			    context = _ref83.context;

			return this.post('/v1/products/' + product + '/team', { username: username }, auth, context);
		}

		/**
   * Remove Particle user to a product team
   * @param  {Object} options Options for this API call
   * @param  {String} options.username  Username for the Particle account
   * @param  {String} options.product Team for this product ID or slug
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'removeTeamMember',
		value: function removeTeamMember(_ref84) {
			var username = _ref84.username,
			    product = _ref84.product,
			    auth = _ref84.auth,
			    context = _ref84.context;

			return this.delete('/v1/products/' + product + '/team/' + username, undefined, auth, context);
		}

		/**
   * Fetch details about a serial number
   * @param  {Object} options Options for this API call
   * @param  {String} options.serialNumber The serial number printed on the barcode of the device packaging
   * @param  {String} options.auth Access Token
   * @returns {Promise} A promise
   */

	}, {
		key: 'lookupSerialNumber',
		value: function lookupSerialNumber(_ref85) {
			var serialNumber = _ref85.serialNumber,
			    auth = _ref85.auth,
			    context = _ref85.context;

			return this.get('/v1/serial_numbers/' + serialNumber, auth, undefined, context);
		}

		/**
   * Create a mesh network
   * @param {Object} options Options for this API call
   * @param {String} options.name Network name
   * @param {String} options.deviceId Gateway device ID
   * @param {String} [options.iccid] ICCID of the active SIM card (only for cellular gateway devices)
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'createMeshNetwork',
		value: function createMeshNetwork(_ref86) {
			var name = _ref86.name,
			    deviceId = _ref86.deviceId,
			    iccid = _ref86.iccid,
			    auth = _ref86.auth,
			    context = _ref86.context;

			return this.post('/v1/networks', { name: name, device_id: deviceId, iccid: iccid }, auth, context);
		}

		/**
   * Remove a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} options.networkId Network ID or name
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'removeMeshNetwork',
		value: function removeMeshNetwork(_ref87) {
			var networkId = _ref87.networkId,
			    auth = _ref87.auth,
			    context = _ref87.context;

			return this.delete('/v1/networks/' + networkId, undefined, auth, context);
		}

		/**
   * List all mesh networks
   * @param {Object} options Options for this API call
   * @param {String} options.auth Access token
   * @param {Number} [options.page] Current page of results
   * @param {Number} [options.perPage] Records per page
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'listMeshNetworks',
		value: function listMeshNetworks(_ref88) {
			var auth = _ref88.auth,
			    context = _ref88.context,
			    page = _ref88.page,
			    perPage = _ref88.perPage;

			var query = page ? { page: page, per_page: perPage } : undefined;
			return this.get('/v1/networks', auth, query, context);
		}

		/**
   * Get information about a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} options.networkId Network ID or name
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'getMeshNetwork',
		value: function getMeshNetwork(_ref89) {
			var networkId = _ref89.networkId,
			    auth = _ref89.auth,
			    context = _ref89.context;

			return this.get('/v1/networks/' + networkId, auth, undefined, context);
		}

		/**
   * Modify a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} options.networkId Network ID or name
   * @param {String} options.action 'add-device', 'remove-device', 'gateway-enable' or 'gateway-disable'
   * @param {String} options.deviceId Device ID
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'updateMeshNetwork',
		value: function updateMeshNetwork(_ref90) {
			var networkId = _ref90.networkId,
			    action = _ref90.action,
			    deviceId = _ref90.deviceId,
			    auth = _ref90.auth,
			    context = _ref90.context;

			return this.put('/v1/networks/' + networkId, { action: action, device_id: deviceId }, auth, context);
		}

		/**
   * Add a device to a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} options.networkId Network ID or name
   * @param {String} options.deviceId Device ID
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'addMeshNetworkDevice',
		value: function addMeshNetworkDevice(_ref91) {
			var networkId = _ref91.networkId,
			    deviceId = _ref91.deviceId,
			    auth = _ref91.auth,
			    context = _ref91.context;

			return this.updateMeshNetwork({ action: 'add-device', networkId: networkId, deviceId: deviceId, auth: auth, context: context });
		}

		/**
   * Remove a device from a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} [options.networkId] Network ID or name
   * @param {String} options.deviceId Device ID
   * @param {String} options.auth Access token
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'removeMeshNetworkDevice',
		value: function removeMeshNetworkDevice(_ref92) {
			var networkId = _ref92.networkId,
			    deviceId = _ref92.deviceId,
			    auth = _ref92.auth,
			    context = _ref92.context;

			if (!networkId) {
				return this.delete('/v1/devices/' + deviceId + '/network', undefined, auth, context);
			}
			return this.updateMeshNetwork({ action: 'remove-device', networkId: networkId, deviceId: deviceId, auth: auth, context: context });
		}

		/**
   * List all devices of a mesh network.
   * @param {Object} options Options for this API call
   * @param {String} options.networkId Network ID or name
   * @param {String} options.auth Access token
   * @param {Number} [options.role] Device role: 'gateway' or 'node'
   * @param {Number} [options.page] Current page of results
   * @param {Number} [options.perPage] Records per page
   * @param {Object} [options.context] Request context
   * @returns {Promise<Object>} A promise
   */

	}, {
		key: 'listMeshNetworkDevices',
		value: function listMeshNetworkDevices(_ref93) {
			var networkId = _ref93.networkId,
			    auth = _ref93.auth,
			    role = _ref93.role,
			    page = _ref93.page,
			    perPage = _ref93.perPage,
			    context = _ref93.context;

			var query = role || page ? { role: role, page: page, per_page: perPage } : undefined;
			return this.get('/v1/networks/' + networkId + '/devices', auth, query, context);
		}

		/**
   * API URI to access a device
   * @param  {Object} options Options for this API call
   * @param  {String} options.deviceId  Device ID to access
   * @param  {String} [options.product] Device only in this product ID or slug
   * @private
   * @returns {string} URI
   */

	}, {
		key: 'deviceUri',
		value: function deviceUri(_ref94) {
			var deviceId = _ref94.deviceId,
			    product = _ref94.product;

			return product ? '/v1/products/' + product + '/devices/' + deviceId : '/v1/devices/' + deviceId;
		}
	}, {
		key: 'get',
		value: function get(uri, auth, query, context) {
			context = this._buildContext(context);
			return this.agent.get(uri, auth, query, context);
		}
	}, {
		key: 'head',
		value: function head(uri, auth, query, context) {
			context = this._buildContext(context);
			return this.agent.head(uri, auth, query, context);
		}
	}, {
		key: 'post',
		value: function post(uri, data, auth, context) {
			context = this._buildContext(context);
			return this.agent.post(uri, data, auth, context);
		}
	}, {
		key: 'put',
		value: function put(uri, data, auth, context) {
			context = this._buildContext(context);
			return this.agent.put(uri, data, auth, context);
		}
	}, {
		key: 'delete',
		value: function _delete(uri, data, auth, context) {
			context = this._buildContext(context);
			return this.agent.delete(uri, data, auth, context);
		}
	}, {
		key: 'request',
		value: function request(args) {
			args.context = this._buildContext(args.context);
			return this.agent.request(args);
		}
	}, {
		key: 'client',
		value: function client() {
			var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

			return new _Client2.default((0, _assign2.default)({ api: this }, options));
		}
	}]);
	return Particle;
}();

// Aliases for backwards compatibility


Particle.prototype.removeAccessToken = Particle.prototype.deleteAccessToken;

exports.default = Particle;
module.exports = exports['default'];
